zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/http/ElasticsearchHttpSpanConsumer.java

```diff
@@ -37,29 +38,95 @@
+    Callback<Void> callbackWrapper = new Callback<Void>() {
+        @Override
+        public void onSuccess(Void value) {
+            callback.onSuccess(value);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            clearIndexToServiceSpansCache(indexToServiceSpans, indexToServiceSpansCache);
+            callback.onError(t);
+        }
+    };
+
     if (spans.isEmpty()) {
-      callback.onSuccess(null);
+      callbackWrapper.onSuccess(null);
       return;
     }
     try {
       HttpBulkIndexer indexer = new HttpBulkIndexer("index-span", es);
-      Map<String, Set<Pair<String>>> indexToServiceSpans = indexSpans(indexer, spans);
+      indexToServiceSpans = indexSpans(indexer, spans);
+
+      // remove already cached items from indexToServiceSpans, add new items to cache
+      for (Map.Entry<String, Set<Pair<String>>> entry : indexToServiceSpans.entrySet()) {
+        Set<Pair<String>> serviceSpansEntryCached = null;
+        synchronized(indexToServiceSpansCache) {
+          serviceSpansEntryCached = indexToServiceSpansCache.get(entry.getKey());
+        }
+        if (serviceSpansEntryCached == null) {
+          synchronized(indexToServiceSpansCache) {
+            indexToServiceSpansCache.put(entry.getKey(), serviceSpansEntryCached = Collections.newSetFromMap(new LinkedHashMap<Pair<String>, Boolean>() {
+              @Override
+              protected boolean removeEldestEntry(Map.Entry<Pair<String>, Boolean> eldest) {
+                return size() > MAX_CACHE_ENTRIES;
+              }
+            }));
+          }
+        }
+        Set<Pair<String>> serviceSpansEntry = entry.getValue();
+        synchronized(serviceSpansEntryCached) {
+          serviceSpansEntry.removeAll(serviceSpansEntryCached);
+          serviceSpansEntryCached.addAll(serviceSpansEntry);
+        }
+      }
+
       if (!indexToServiceSpans.isEmpty()) {
         indexNames(indexer, indexToServiceSpans);
       }
-      indexer.execute(callback);
+      indexer.execute(callbackWrapper);
     } catch (Throwable t) {
       propagateIfFatal(t);
-      callback.onError(t);
+      callbackWrapper.onError(t);
     }
   }
+  /** Remove recently added items from servicespan cache */
+  void clearIndexToServiceSpansCache(Map<String, Set<Pair<String>>> index, Map<String, Set<Pair<String>>> cache) {
+    for (Map.Entry<String, Set<Pair<String>>> entry : index.entrySet()) {
+      Set<Pair<String>> serviceSpansEntryCached = null;
+      synchronized (cache) {
+        serviceSpansEntryCached = cache.get(entry.getKey());
+      }
+      if (serviceSpansEntryCached == null || serviceSpansEntryCached.isEmpty()) continue;
+
+      synchronized (serviceSpansEntryCached) {
+        serviceSpansEntryCached.removeAll(entry.getValue());
+      }
+    }
+  }
+
+  /** Clear servicespan cache (for testing purposes) */
+  void resetIndexToServiceSpansCache() {
+    indexToServiceSpansCache.clear();
```

Comment:
this should be synchronized on indexToServiceSpansCache just to make it obvious. I'd probably make a package-private type that encapsulates access to this map, with the map itself private. All things that affect it would be package protected synchronized methods to make it easier to reason with.